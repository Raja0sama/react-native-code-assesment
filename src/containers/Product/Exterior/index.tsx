import {View, Text, Image, TouchableOpacity} from 'react-native';
import React, {useMemo, useState} from 'react';
import Black from '@assets/black.svg';
import Blue from '@assets/blue.svg';
import Gray from '@assets/gray.svg';
import Red from '@assets/red.svg';
import White from '@assets/white.svg';
import {CarData} from '@global/reducers/cars';
import {BottomSection} from '@components/BottomSection';

const newColorsObject = [
  {
    IconComponent: White,
    label: 'Pearl White',
    image:
      'https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PPSW,$W40B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=1&crop=1300,500,300,300&version=v0121-459f1b30d202212210637',
  },
  {
    IconComponent: Gray,
    label: 'Midnight Silver',
    image:
      'https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PMNG,$W40B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=1&crop=1300,500,300,300&version=v0121-459f1b30d202212210637',
  },
  {
    IconComponent: Blue,
    label: 'Deep Blue',
    image:
      'https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PPSB,$W40B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=1&crop=1300,500,300,300&version=v0121-459f1b30d202212210637',
  },
  {
    IconComponent: Red,
    label: 'red',
    image:
      'https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PPMR,$W40B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=1&crop=1300,500,300,300&version=v0121-459f1b30d202212210637',
  },
  {
    IconComponent: Black,
    label: 'Red Multi-Coat',
    image:
      'https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PMNG,$W40B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=1&crop=1300,500,300,300&version=v0121-459f1b30d202212210637',
  },
];

interface CarProps {
  route: {
    params: {carDetail: CarData};
  };
}

export default function Exterior(props: CarProps) {
  const {carDetail} = props.route.params;

  const [selectedCar, setCar] = useState<typeof newColorsObject[0]>(
    newColorsObject[0],
  );

  const getPrice = useMemo(() => {
    const basePrice = parseInt(carDetail.basePrice.split(',').join(''), 10);
    const item = carDetail.exteriorColors.find(
      e => e.name === selectedCar.label,
    );
    if (!item) {
      return basePrice;
    }

    if (item.price === 'Included') {
      return basePrice;
    }

    const price = parseInt((item.price as string).split(',').join(''), 10);
    return price + basePrice;
  }, [carDetail.basePrice, carDetail.exteriorColors, selectedCar.label]);

  const selectedColor = {
    borderColor: 'red',
    borderWidth: 5,
    borderRadius: 100,
  };

  return (
    <View className="flex flex-1 bg-[#f2f5f8]">
      <View className="px-9 py-9 flex-1 flex-col justify-between">
        <Text className="pb-10  text-[#A4B0BC] text-xl font-[Nunito-Regular]">
          Select Color
        </Text>
        <View className="items-center flex-1 ">
          <Image
            resizeMode="contain"
            className="w-screen h-full"
            source={{uri: selectedCar.image}}
          />
        </View>
        <View className="flex-1 mt-5">
          <Text className=" font-[Nunito-Regular] text-black text-2xl">
            Pearl White Multi-Coat
          </Text>
          <Text className=" font-[Nunito-Regular] text-[#D01000] text-xl">
            Included
          </Text>
          <View className="flex-row justify-between items-center pt-5">
            {newColorsObject.map(color => {
              const {IconComponent, label} = color;
              return (
                <TouchableOpacity
                  key={label}
                  onPress={() => setCar(color)}
                  style={selectedCar.label === label && selectedColor}>
                  <IconComponent />
                </TouchableOpacity>
              );
            })}
          </View>
          <View className="h-[1px] w-full bg-[#CFD3D9] mt-4" />
        </View>
        <View className="flex-1 justify-center">
          <Text className=" font-[Nunito-Regular] text-black pb-5">
            20’’ Performance Wheels
          </Text>
          <Text className=" font-[Nunito-Regular] text-black ">
            Carbon fiber spoiler
          </Text>
        </View>
      </View>
      <View className="">
        <View className=" bg-white py-10 px-9 rounded-t-[36px]">
          <BottomSection navigateTo="INTERIOR" basePrice={getPrice + ''} />
        </View>
      </View>
    </View>
  );
}
