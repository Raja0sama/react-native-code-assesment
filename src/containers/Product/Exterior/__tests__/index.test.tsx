import React from 'react';
import Exterior from '../index';
import renderer from 'react-test-renderer';

jest.mock('../../../../assets/black.svg', () => '<View />');
jest.mock('../../../../assets/blue.svg', () => '<View />');
jest.mock('../../../../assets/gray.svg', () => '<View />');
jest.mock('../../../../assets/red.svg', () => '<View />');
jest.mock('../../../../assets/white.svg', () => '<View />');

jest.mock('@react-navigation/native');

describe('Exterior', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Exterior
          route={{
            params: {
              carDetail: response,
            },
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const response = {
  id: 1,
  model: 'Model 3',
  basePrice: '184,000',
  features: [
    {
      id: 1,
      name: 'Performance',
      price: '235,000',
    },
    {
      id: 2,
      name: 'Long Range',
      price: '215,000',
    },
  ],
  exteriorColors: [
    {
      id: 1,
      name: 'Pearl White',
      price: 'Included',
    },
    {
      id: 5,
      name: 'Solid Black',
      price: '4500',
    },
    {
      id: 9,
      name: 'Midnight Silver',
      price: '6000',
    },
    {
      id: 13,
      name: 'Deep Blue',
      price: '6000',
    },
    {
      id: 17,
      name: 'Red Multi-Coat',
      price: '7500',
    },
  ],
  interiorOptions: [
    {
      id: 1,
      name: 'All Black',
      price: 'Included',
    },
    {
      id: 5,
      name: 'Black & White',
      price: '4,500',
    },
  ],
  autopilotOptions: [
    {
      id: 1,
      name: 'Autopilot',
      price: '14,000',
    },
    {
      id: 5,
      name: 'Full self-driving',
      price: '28,000',
    },
  ],
  imageLink:
    'https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PPMR,$W41B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=2&crop=0,0,0,0&version=v0121-078c13b0d202212151058',
};
