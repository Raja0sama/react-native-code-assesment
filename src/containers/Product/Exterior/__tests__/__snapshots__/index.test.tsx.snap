// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exterior renders correctly 1`] = `
<View
  style={
    [
      {
        "display": "flex",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "backgroundColor": "#f2f5f8",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "paddingLeft": 36,
          "paddingRight": 36,
        },
        {
          "paddingBottom": 36,
          "paddingTop": 36,
        },
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "flexDirection": "column",
        },
        {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "paddingBottom": 40,
          },
          {
            "color": "#A4B0BC",
          },
          {
            "fontSize": 20,
            "lineHeight": 28,
          },
          {
            "fontFamily": "Nunito-Regular",
          },
        ]
      }
    >
      Select Color
    </Text>
    <View
      style={
        [
          {
            "alignItems": "center",
          },
          {
            "flexBasis": "0%",
            "flexGrow": 1,
            "flexShrink": 1,
          },
        ]
      }
    >
      <Image
        resizeMode="contain"
        source={
          {
            "uri": "https://static-assets.tesla.com/configurator/compositor?context=design_studio_2&options=$MT322,$PPSW,$W40B,$IBB1&view=STUD_FRONT34&model=m3&size=1920&bkba_opt=1&crop=1300,500,300,300&version=v0121-459f1b30d202212210637",
          }
        }
        style={
          [
            {
              "width": 750,
            },
            {
              "height": "100%",
            },
          ]
        }
      />
    </View>
    <View
      style={
        [
          {
            "flexBasis": "0%",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          {
            "marginTop": 20,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Nunito-Regular",
            },
            {
              "color": "#000",
            },
            {
              "fontSize": 24,
              "lineHeight": 32,
            },
          ]
        }
      >
        Pearl White Multi-Coat
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "Nunito-Regular",
            },
            {
              "color": "#D01000",
            },
            {
              "fontSize": 20,
              "lineHeight": 28,
            },
          ]
        }
      >
        Included
      </Text>
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "justifyContent": "space-between",
            },
            {
              "alignItems": "center",
            },
            {
              "paddingTop": 20,
            },
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "borderColor": "red",
              "borderRadius": 100,
              "borderWidth": 5,
              "opacity": 1,
            }
          }
        >
          <<View /> />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <<View /> />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <<View /> />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <<View /> />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <<View /> />
        </View>
      </View>
      <View
        style={
          [
            {
              "height": 1,
            },
            {
              "width": "100%",
            },
            {
              "backgroundColor": "#CFD3D9",
            },
            {
              "marginTop": 16,
            },
          ]
        }
      />
    </View>
    <View
      style={
        [
          {
            "flexBasis": "0%",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          {
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Nunito-Regular",
            },
            {
              "color": "#000",
            },
            {
              "paddingBottom": 20,
            },
          ]
        }
      >
        20’’ Performance Wheels
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "Nunito-Regular",
            },
            {
              "color": "#000",
            },
          ]
        }
      >
        Carbon fiber spoiler
      </Text>
    </View>
  </View>
  <View>
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          {
            "paddingBottom": 40,
            "paddingTop": 40,
          },
          {
            "paddingLeft": 36,
            "paddingRight": 36,
          },
          {
            "borderTopLeftRadius": 36,
            "borderTopRightRadius": 36,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "alignItems": "center",
            },
            {
              "justifyContent": "space-between",
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 24,
                "lineHeight": 32,
              },
              {
                "color": "#000",
              },
              {
                "fontFamily": "Nunito-Regular",
              },
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          AED 
          184000
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#D01000",
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderBottomWidth": 2,
              "borderLeftColor": "#D01000",
              "borderLeftWidth": 2,
              "borderRightColor": "#D01000",
              "borderRightWidth": 2,
              "borderTopColor": "#D01000",
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "borderTopWidth": 2,
              "display": "flex",
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "opacity": 1,
              "paddingBottom": 12,
              "paddingTop": 12,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontFamily": "Nunito-Regular",
                },
                {
                  "fontSize": 20,
                  "lineHeight": 28,
                },
                {
                  "color": "#000",
                },
              ]
            }
          >
            Next
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
